src/CoreFoundation/CFProxySupport.cs

```diff
@@ -579,62 +581,328 @@ public static CFProxySettings GetSystemProxySettings ()
+				CFObject.CFRetain (proxyList);
+				pacCbData.ProxyListPtr = proxyList;
+			}
+			pacCbData.ErrorPtr = error;
+			// stop the CFRunLoop
+			var runLoop = new CFRunLoop (pacCbData.CFRunLoopPtr);
+			Marshal.StructureToPtr (pacCbData, client, false);
+			runLoop.Stop ();
+		}
+
+		static async Task<(CFProxy[] proxies, NSError error)> ExecutePacCFRunLoopSourceAsync (CreatePACCFRunLoopSource factory, CancellationToken cancellationToken)
+		{
+			CFProxy[] proxies = null;
+			NSError outError = null;
+			await Task.Run (() => {
+				// we need the runloop of THIS thread, so it is important to get it in the correct context
+				var runLoop = CFRunLoop.Current;
+
+				// build a struct that will have all the needed info for the callback
+				var pacCbData = new PACProxyCallbackData ();
+				pacCbData.CFRunLoopPtr = runLoop.Handle;
+				var pacDataPtr = Marshal.AllocHGlobal (Marshal.SizeOf(pacCbData));
+				try {
+					Marshal.StructureToPtr (pacCbData, pacDataPtr, false); 
+
+					var clientContext = new CFStreamClientContext ();
+					clientContext.Info = pacDataPtr;
+
+					using (var loopSource = new CFRunLoopSource (factory (ExecutePacCallback, ref clientContext)))
+					using (var mode = new NSString ("Xamarin.iOS.Proxy")) {
+						runLoop.AddSource (loopSource, mode);
+						// blocks until stop is called, will be done in the cb set previously
+						runLoop.RunInMode (mode, double.MaxValue, false);
+						runLoop.RemoveSource (loopSource, mode);
+					}
+					pacCbData = (PACProxyCallbackData) Marshal.PtrToStructure(pacDataPtr, typeof (PACProxyCallbackData));
+					// get data from the struct
+					proxies = pacCbData.ProxyList;
+					outError = pacCbData.Error;
+				} finally {
+					// clean resources
+					Marshal.FreeHGlobal (pacDataPtr); 
+				}
+			}, cancellationToken).ConfigureAwait (false);
+
+			return (proxies: proxies, error: outError);
+		}
+
+		static CFProxy[] ExecutePacCFRunLoopSourceBlocking (CreatePACCFRunLoopSource factory, out NSError outError)
+		{
+			var runLoop = CFRunLoop.Current;
+			CFProxy [] proxies = null;
+			outError = null;
+
+			// build a struct that will have all the needed info for the callback
+			var pacCbData = new PACProxyCallbackData ();
+			pacCbData.CFRunLoopPtr = runLoop.Handle;
+			var pacDataPtr = Marshal.AllocHGlobal (Marshal.SizeOf(pacCbData));
+			try {
+				Marshal.StructureToPtr (pacCbData, pacDataPtr, false); 
+
+				var clientContext = new CFStreamClientContext ();
+				clientContext.Info = pacDataPtr;
+
+				using (var loopSource = new CFRunLoopSource (factory (ExecutePacCallback, ref clientContext)))
+				using (var mode = new NSString ("Xamarin.iOS.Proxy")) {
+					runLoop.AddSource (loopSource, mode);
+					runLoop.RunInMode (mode, double.MaxValue, false);
+					runLoop.RemoveSource (loopSource, mode);
+				}
+				pacCbData = (PACProxyCallbackData) Marshal.PtrToStructure(pacDataPtr, typeof (PACProxyCallbackData));
+				// get data from the struct
+				if (pacCbData.Error != null)
```

Comment:
Can't you just get both values in both cases?

```csharp
outError = pacCbData.Error;
return pacCbData.ProxyList;
```

it's a few lines shorter :smile: